django/contrib/admin/views/main.py

```diff
@@ -135,6 +138,33 @@ def get_filters(self, request):
             if spec and spec.has_output():
                 filter_specs.append(spec)
 
+        if self.date_hierarchy:
+            year = lookup_params.pop('{}__year'.format(self.date_hierarchy), None)
+            if year is not None:
+                month = lookup_params.pop('{}__month'.format(self.date_hierarchy), None)
+                day = lookup_params.pop('{}__day'.format(self.date_hierarchy), None)
+
+                from_date = datetime.datetime(
+                    int(year),
+                    int(month if month is not None else 1),
+                    int(day if day is not None else 1),
+                )
+                if settings.USE_TZ:
+                    from_date = get_default_timezone().localize(from_date)
+
+                if day:
+                    to_date = from_date + datetime.timedelta(days=1)
+
+                elif month:
+                    assert from_date.day == 1
```

Comment:
You can drop this assertion, the way `from_date` is constructed and that this branch is behind the `if day` one makes it impossible to reach.